spmd/tensor/parallel/multihead_attention_tp.py

```diff
@@ -0,0 +1,254 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+
+import math
+
+import torch
+from spmd.tensor import DTensor as DT
+from spmd.tensor.placement_types import Shard
+
+
+def _view_with_sharding_dim_change(tensor, sharding_dim, shape):
+    """
+    Sometimes we want to change the implicit sharding dim for a 
+    distributed tensor without comms. 
+    """
+    if isinstance(tensor, DT):
+        return tensor._view_with_sharding_dim_change(sharding_dim, shape)
+    else:
+        return tensor.view(shape)
+
+
+def _stride_same_as_shard(tensor, tp_size, chunk_dim, cat_dim):
+    """
+    Adjust local tensor's stride same as the sharded situation.
+    So that view result will keeps the same.
+    """
+    if isinstance(tensor, DT):
+        return tensor
+    view_size = list(tensor.size())
+    view_size[chunk_dim] //= tp_size
+    return torch.cat(
+        [
+            t.view(*view_size)
+            for t in tensor.chunk(tp_size, dim=chunk_dim)
+        ],
+        dim=cat_dim,
+    ).contiguous()
+
+
+class TensorParallelMultiheadAttention(torch.nn.Module):
+    """
+    Multi-head Attention block from Transformer models.
+    Since we need some customizations for the attention layer,
+    we are writing a customized but mathematically equivalent
+    attention module as defined in torch.nn.
+
+    Note that:
+    We now only support the case when it's self attention with
+    limited input args and we also assume that the input tensor
+    has a dimension of three. Although we do implement the logic
+    for multihead attention, it was not fully tested.
```

Comment:
Do we need to extend this test to cover the case in future?

Reply:
Yep, we will add more tests down the road as we implemented more features for Multihead Attention.